doc/sphinx/proofs/automatic-tactics/auto.rst

```diff
@@ -431,16 +526,21 @@ Creating Hints
          hint_info ::= %| {? @natural } {? @one_pattern }
          one_pattern ::= @one_term
 
-      The first form adds each :n:`@qualid` as a hint with the head symbol of the type of
-      :n:`@qualid` to the specified hint databases (:n:`@ident`\s). The cost of the hint is the number of
-      subgoals generated by :tacn:`simple apply` :n:`@qualid` or, if specified, :n:`@natural`. The
+      The first form adds each :n:`@qualid` as a hint with the head symbol of the
+      type of :n:`@qualid` to the specified hint databases (:n:`@ident`\s). The
+      cost of the hint is :n:`@natural` (which can be inside :n:`@hint_info`). The
       associated pattern is inferred from the conclusion of the type of
       :n:`@qualid` or, if specified, the given :n:`@one_pattern`.
 
       If the inferred type
-      of :n:`@qualid` does not start with a product, :tacn:`exact` :n:`@qualid` is added
-      to the hint list.  If the type can be reduced to a type starting with a product,
-      :tacn:`simple apply` :n:`@qualid` is also added to the hints list.
+      of :n:`@qualid` does not start with a product, the command adds :tacn:`exact`
+      :n:`@qualid` to the hint list.  If the inferred type starts with a product,
+      the command adds :tacn:`simple apply` :n:`@qualid` to the hints list.  Note that
+      unlike :tacn:`auto` and :tacn:`eauto`,
```

Comment:
Being reducable to a product and starting with a product is not the same. In once case, there is conversion / or hnf involved, in the other it is just a syntactic match. What is going on here ?

Reply:
This is determined when you add a hint (`Hint Resolve`), not when you're executing auto, et al.  For example:

```
Theorem x: forall a:nat,  a = a.
Admitted.
Theorem y: 1 = 2.
Admitted.

Create HintDb foo.
Hint Resolve x : foo.
Hint Resolve y : foo.
Print HintDb foo.

(*
Non-discriminated database
Unfoldable variable definitions: all
Unfoldable constant definitions: all
Unfoldable projection definitions: all
Cut: emp
For any goal ->   
For eq ->   exact y (level 0, pattern 1 = 2, id 0)
            simple apply x (level 0, pattern ?M151 = ?M151, id 0)
*)

Can you give an example of something that's reducible to a product?  We can try that, too, but I'm pretty sure `Hint Resolve` is doing any reduction.